### ISSUE TITLE
[RyuJIT/ARM32] Assertion failed 'size >= sizeof(int)'

### ISSUE THREAD

**Issue by jyoungyun** [dotnet/coreclr#12990]
# Description
Assertion:
```
Assert failure(PID 23320 [0x00005b18], Thread: 23320 [0x5b18]): Assertion failed 'size >= sizeof(int)' in 'overldrem:Main():int' (IL size 28941)

    File: /opt/code/src/jit/regset.cpp Line: 3348
    Image: /home/pi3/Downloads/ryujit/overlay-0718-fixed3/corerun
```

# Failed case
```
JIT/Methodical/divrem/div/overlddiv_cs_do/overlddiv_cs_do.sh
JIT/Methodical/divrem/div/overlddiv_cs_ro/overlddiv_cs_ro.sh
JIT/Methodical/divrem/rem/overldrem_cs_do/overldrem_cs_do.sh
JIT/Methodical/divrem/rem/overldrem_cs_ro/overldrem_cs_ro.sh
```

**Comment by jyoungyun** [comment]
@BruceForstall Please assign to me.

**Comment by jyoungyun** [comment]
This is a call stack.
```
#13 0x7136b1aa in Compiler::tmpPreAllocateTemps (this=0x86918, type=var_types::TYP_STRUCT, count=1) at /opt/code/src/jit/regset.cpp:3348
#14 0x7131b5ac in LinearScan::recordMaxSpill (this=0x9923c8) at /opt/code/src/jit/lsra.cpp:8551
#15 0x71311f32 in LinearScan::resolveRegisters (this=0x9923c8) at /opt/code/src/jit/lsra.cpp:9164
#16 0x7130d8a6 in LinearScan::doLinearScan (this=0x9923c8) at /opt/code/src/jit/lsra.cpp:1778
#17 0x712296ce in Compiler::compCompile (this=0x86918, methodCodePtr=0x7effc874, methodCodeSize=0x7effcd18, compileFlags=0x7effc888)
    at /opt/code/src/jit/compiler.cpp:4829
```

After adding below codes:
```
diff --git a/src/jit/lsra.cpp b/src/jit/lsra.cpp
index 28617b9..191d037 100644
--- a/src/jit/lsra.cpp
+++ b/src/jit/lsra.cpp
@@ -8613,6 +8613,12 @@ void LinearScan::updateMaxSpill(RefPosition* refPosition)
                     ReturnTypeDesc* retTypeDesc = treeNode->AsCall()->GetReturnTypeDesc();
                     typ                         = retTypeDesc->GetReturnRegType(refPosition->getMultiRegIdx());
                 }
+#ifdef _TARGET_ARM_
+                else if (treeNode->OperIsPutArgSplit())
+                {
+                    typ = treeNode->AsPutArgSplit()->GetRegType(refPosition->getMultiRegIdx());
+                }
+#endif
                 else
```

New assertion generated:
```
Assert failure(PID 5095 [0x000013e7], Thread: 5095 [0x13e7]): Assertion failed '!"gtArgEntryByNode: node not found"' in 'overldrem:Main():int' (IL size 28941)

    File: /opt/code/src/jit/gentree.cpp Line: 6488
```

I think this assertion is not related to code changes.

Original gen tree:
```
                                                  /--*  t3351  int
                                                  +--*  t3367  int
                                                  +--*  t3383  int
                                                  +--*  t3399  int
N066 (187,131) [003352] ---XG--------     t3352 = *  FIELD_LIST void   int at offset 0 <l:$cac, c:$cad>
                                                  /--*  t5196  byref
N079 ( 46, 31) [002485] ---X---------     t2485 = *  IND       ref    <l:$36c, c:$baf>
                                                  /--*  t3352  void   arg1 r2,r3 out+00
                                                  +--*  t2485  ref    arg0 in r0
N082 (249,174) [002499] --CXG--------     t2499 = *  CALL      int    numHolder.op_Division $c35
```
But after some logic is executed:
```
                                                  /--*  t3351  int
                                                  +--*  t3367  int
                                                  +--*  t3383  int
                                                  +--*  t3399  int
N066 (187,131) [003352] ---XG--------     t3352 = *  FIELD_LIST void   int at offset 0 <l:$cac, c:$cad>
                                                  /--*  t3352  void
               [005197] ---XG--------     t5197 = *  PUTARG_SPLIT struct
                                                  /--*  t5196  byref
N079 ( 46, 31) [002485] ---X---------     t2485 = *  IND       ref    <l:$36c, c:$baf>
                                                  /--*  t2485  ref
               [005198] ---X---------     t5198 = *  PUTARG_REG ref
                                                  /--*  t5197  struct arg1 r2,r3 out+00
                                                  +--*  t5198  ref    arg0 in r0
                                                  +--*  t5200  int    control expr
N082 (249,174) [002499] --CXG--------     t2499 = *  CALL      int    numHolder.op_Division $c35
```
After doing lsra, t2499 has a t6841 as a child instead of t5197. And t6841 is GT_RELOAD node.
GT_RELOAD means that it is reloaded to a different register.
```
                                                  /--*  t3351  int
                                                  +--*  t3367  int
                                                  +--*  t6840  int
                                                  +--*  t3399  int
N5013 (187,131) [003352] -c-XG--------     t3352 = *  FIELD_LIST void   int at offset 0 REG NA <l:$cac, c:$cad>
                                                  /--*  t3352  void
N5015 (???,???) [005197] ---XG-------Z     t5197 = *  PUTARG_SPLIT struct REG r2
                                                  /--*  t5197  struct
               [006841] ---XG--------     t6841 = *  RELOAD    struct REG lr
                                                  /--*  t5196  byref
N5095 ( 46, 31) [002485] ---X---------     t2485 = *  IND       ref    REG r0 <l:$36c, c:$baf>
                                                  /--*  t2485  ref
N5097 (???,???) [005198] ---X---------     t5198 = *  PUTARG_REG ref    REG r0
                                                  /--*  t6841  struct arg1 r2,r3 out+00
                                                  +--*  t5198  ref    arg0 in r0
                                                  +--*  t5200  int    control expr
N5103 (249,174) [002499] --CXG--------     t2499 = *  CALL      int    numHolder.op_Division $c35
```
And here is code generation:
```
                                                              /--*  t3351  int
                                                              +--*  t3367  int
                                                              +--*  t6840  int
                                                              +--*  t3399  int
Generating: N5013 (187,131) [003352] -c-XG--------     t3352 = *  FIELD_LIST void   int at offset 0 REG NA <l:$cac, c:$cad>
                                                              /--*  t3352  void
Generating: N5015 (???,???) [005197] ---XG-------Z     t5197 = *  PUTARG_SPLIT struct REG r2
                                                        Tree-Node marked unspilled from  [003351]
IN099a:             ldr     r2, [sp+0x1bc]      // [TEMP_02]
release temp #2, slot 0, size = 4
                                                        Tree-Node marked unspilled from  [003383]
IN099b:             ldr     lr, [sp+0x1b8]      // [TEMP_01]
release temp #1, slot 0, size = 4

      G_M13827_IG25:        ; offs=001C58H, funclet=00
IN099c:             str     lr, [sp]    // [V110 OutArgs]
IN099d:             str     r0, [sp+0x04]       // [V110 OutArgs+0x04]
reused temp #1, slot 0, size = 4
                                                        The register r2 spilled with    [005197]
IN099e:             str     r2, [sp+0x1b8]      // [TEMP_01]
                                                              /--*  t5197  struct
Generating:                [006841] ---XG--------     t6841 = *  RELOAD    struct REG lr
                                                              /--*  t5196  byref
Generating: N5095 ( 46, 31) [002485] ---X---------     t2485 = *  IND       ref    REG r0 <l:$36c, c:$baf>
IN09cc:             add     r2, r5, r0 LSL 2
IN09cd:             ldr     r0, [r2+24]
                                                        GC regs: 03E0 {r5 r6 r7 r8 r9} => 03E1 {r0 r5 r6 r7 r8 r9}
                                                              /--*  t2485  ref
Generating: N5097 (???,???) [005198] ---X---------     t5198 = *  PUTARG_REG ref    REG r0
                                                        GC regs: 03E1 {r0 r5 r6 r7 r8 r9} => 03E0 {r5 r6 r7 r8 r9}
                                                        GC regs: 03E0 {r5 r6 r7 r8 r9} => 03E1 {r0 r5 r6 r7 r8 r9}
                                                              /--*  t6841  struct arg1 r2,r3 out+00
                                                              +--*  t5198  ref    arg0 in r0
                                                              +--*  t5200  int    control expr
Generating: N5103 (249,174) [002499] --CXG--------     t2499 = *  CALL      int    numHolder.op_Division $c35
<assertion here>
```

Here is a stack trace:
```
#13 0x712804cc in Compiler::gtArgEntryByNode (call=0xf4398, node=0x17d6a4) at /opt/code/src/jit/gentree.cpp:6488
#14 0x7139d154 in CodeGen::genCallInstruction (this=0x86d80, call=0xf4398) at /opt/code/src/jit/codegenarmarch.cpp:1800
#15 0x7139a424 in CodeGen::genCodeForTreeNode (this=0x86d80, treeNode=0xf4398) at /opt/code/src/jit/codegenarmarch.cpp:286
#16 0x7121face in CodeGen::genCodeForBBlist (this=0x86d80) at /opt/code/src/jit/codegenlinear.cpp:388
#17 0x7120f598 in CodeGen::genGenerateCode (this=0x86d80, codePtr=0x7effc8a4, nativeSizeOfCode=0x7effcd48) at /opt/code/src/jit/codegencommon.cpp:3144
```

Finally, I found that where GT_RELOAD gentree generates.
https://github.com/dotnet/coreclr/blob/master/src/jit/lsra.cpp#L8951-L8952

If the gentree is GT_RELOAD, I think somewhere should assign a valid reg for those positions. Could it be that this part is missing in codegen stage?

The part of assiging valid reg for GT_RELOAD/GT_COPY:
1. genMultiRegCallStoreToLocal
    But this function only works in GT_STORE_LCL_VAR and type is LONG or STRUCT.
2. genStructReturn
3. genUnspillRegIfNeeded (My feeling is that this part is a problem...)

**genUnspillRegIfNeeded**
call stack
```
#0  CodeGen::genUnspillRegIfNeeded (this=0x86d80, tree=0x8a6b4) at /opt/code/src/jit/codegenlinear.cpp:856
#1  0x71220308 in CodeGen::genConsumeReg (this=0x86d80, tree=0x8a6b4) at /opt/code/src/jit/codegenlinear.cpp:1179
#2  0x712210c4 in CodeGen::genConsumeRegs (this=0x86d80, tree=0x8a6b4) at /opt/code/src/jit/codegenlinear.cpp:1288
#3  0x713a2d60 in CodeGen::genCodeForStoreLclVar (this=0x86d80, tree=0x8a77c) at /opt/code/src/jit/codegenarm.cpp:1113
#4  0x7139a2e6 in CodeGen::genCodeForTreeNode (this=0x86d80, treeNode=0x8a77c) at /opt/code/src/jit/codegenarmarch.cpp:175
#5  0x7121face in CodeGen::genCodeForBBlist (this=0x86d80) at /opt/code/src/jit/codegenlinear.cpp:388
#6  0x7120f598 in CodeGen::genGenerateCode (this=0x86d80, codePtr=0x7effc8a4, nativeSizeOfCode=0x7effcd48) at /opt/code/src/jit/codegencommon.cpp:3144
#7  0x7122972a in Compiler::compCompile (this=0x84dd0, methodCodePtr=0x7effc8a4, methodCodeSize=0x7effcd48, compileFlags=0x7effc8b8)
    at /opt/code/src/jit/compiler.cpp:4850
```

```
#0  CodeGen::genUnspillRegIfNeeded (this=0x86d80, tree=0x8a910) at /opt/code/src/jit/codegenlinear.cpp:856
#1  0x71220308 in CodeGen::genConsumeReg (this=0x86d80, tree=0x8a910) at /opt/code/src/jit/codegenlinear.cpp:1179
#2  0x7139c9be in CodeGen::genPutArgReg (this=0x86d80, tree=0x1560f4) at /opt/code/src/jit/codegenarmarch.cpp:898
#3  0x7139a3f8 in CodeGen::genCodeForTreeNode (this=0x86d80, treeNode=0x1560f4) at /opt/code/src/jit/codegenarmarch.cpp:276
#4  0x7121face in CodeGen::genCodeForBBlist (this=0x86d80) at /opt/code/src/jit/codegenlinear.cpp:388
#5  0x7120f598 in CodeGen::genGenerateCode (this=0x86d80, codePtr=0x7effc8a4, nativeSizeOfCode=0x7effcd48) at /opt/code/src/jit/codegencommon.cpp:3144
#6  0x7122972a in Compiler::compCompile (this=0x84dd0, methodCodePtr=0x7effc8a4, methodCodeSize=0x7effcd48, compileFlags=0x7effc8b8)
    at /opt/code/src/jit/compiler.cpp:4850
```

**Comment by hqueue** [comment]
This was fixed by #13190. 
Please close this :)
